using messenger HandleTrait as QueryBus with appropriate result typing

bnowak opened the pull request:
This PR goal is to determine correct result typing for QueryBus kind of classes which uses SF messenger HandleTrait internally and simply return its results.

Commit:
test-cases for using messenger HandleTrait as QueryBus

michaljusiega commented on tests/Type/Symfony/data/messenger_handle_trait.php:
```diff
@@ -111,3 +111,32 @@ public function __invoke()
 }
+
+class QueryBus {
+    use HandleTrait;
+
+    public function dispatch(object $query): mixed
+    {
+        return $this->handle($query);
+    }
```
We could also add this test case with changed function name.

```suggestion
    use HandleTrait {
        handle as private handleQuery;
    }

    public function handle(object $query): mixed
    {
        return $this->handleQuery($query);
    }
```

michaljusiega commented:
@ondrejmirtes Sorry to disturb you, but I think this is one of the pieces that is missing in this whole puzzle to support. Can we move forward ?

bnowak commented:
Hi @michaljusiega, thanks for your reply. I think we can add it - no problem. However, from my perspective it's more important to address the root issue (missing feature) of the topic (described [here](https://github.com/phpstan/phpstan-symfony/issues/424)). My reasoning is without having this result interpreted on bus class level - this is much less useful in general.

ondrejmirtes commented:
So this QueryBus class is something in your codebase and you want the same type inference that classes using HandleTrait and calling the `handle` method have?

You can simply write a dynamic return type extension for `QueryBus::dispatch()` that delegates the type resolution to `new MethodCall($methodCall->var, 'handle', $methodCall->getArgs())`.

Commit:
implementation

Commit:
implementation for interfaces

Commit:
move test parameters to test extension-test-neon file

Commit:
updated README.md

bnowak commented:
@ondrejmirtes It's ready for review now. You were right about my intentions, however I wanted to make it working for any class that uses `HandleTrait` internally. I didn't want to hardcoded it, so made it parameterized. I've added also support for interfaces. Please review in you free time :pray: 

It's likely to be conflicted with my another [PR](https://github.com/phpstan/phpstan-symfony/pull/448) about splitting tests, so once first of these PR will be merged I can resolve conflicts on another one. :wink:

bnowak commented:
@ondrejmirtes any chances to review & merge it soon, please? :)

ondrejmirtes commented on README.md:
```diff
@@ -168,3 +170,94 @@ Call the new env in your `console-application.php`:
+{
+    public function __invoke(GetProductQuery $query): Product
+    {
+        return $this->productRepository->get($query->productId);
```
Can you please add comments to the code in README about the inferred types with and without the extension? I'm still not sure what it does.

bnowak replied:
Thanks for reply.

So it's about inferring types of query bus classes (from CQRS pattern) or any custom class which internally use messenger's HandleTrait to get and return result of any dispatched message. Without the extension phpstan is not able to recognize that and gives generic return type of bus class (which in fact is some more specified query result). 

So the main difference is calling query bus classes, where based on passed/dispatched message we're getting its correct result type.

I've updated README a bit and the the direct outcome is shown [in the end of it](https://github.com/phpstan/phpstan-symfony/pull/423/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R255-R263) - it shows that we're getting `Product` type instead of generic `mixed`. If that's still not precised and unclear, please advice me what else I could add/adjust. Maybe my describing skills are not so good :sweat_smile:

bnowak replied:
@ondrejmirtes any chances to re-look on this please?

bnowak replied:
@ondrejmirtes ping :pray:

bnowak replied:
@ondrejmirtes ping2 🙏

bnowak replied:
@ondrejmirtes please, could you take a look on this again? and sorry for bothering you so many times

Commit:
updated README.md

Commit:
test class name fix

bnowak commented:
@ondrejmirtes could you review it again please?

VincentLanglet commented on src/Type/Symfony/MessengerHandleTraitWrapperReturnTypeExtension.php:
```diff
@@ -0,0 +1,125 @@
+		$arg = $args[0]->value;
+		$argClassNames = $scope->getType($arg)->getObjectClassNames();
+
+		if (count($argClassNames) === 1) {
```
Instead of a check count === 1.

Can't you do
```
if (count($argClassNames) === 0) {
     return null;
}

$returnTypes = [];
foreach ($argClassNames as $argClassName) {
     $messageMap = $this->getMessageMap();
     $returnType = $messageMap->getTypeForClass($argClassNames[0]);

	if ($returnType === null) {
			return null;
     }
     $returnTypes[] = $returnType;
}

return TypeCombinator::union(...$returnTypes);
```
?

The goal would be to support union like
```
$queryBus = rand(0, 1) ? $productBus : $itemBus;
$productOrItem = $queryBus->dispatch($query); // Returns: Product|Item
```

bnowak replied:
:+1: adjusted

VincentLanglet commented on src/Type/Symfony/MessengerHandleTraitWrapperReturnTypeExtension.php:
```diff
@@ -0,0 +1,125 @@
+		$varType = $scope->getType($expr->var);
+		$classNames = $varType->getObjectClassNames();
+
+		if (count($classNames) !== 1) {
```
Can't you do something like
```
if (count($classNames) === 0) {
     return false;
}

foreach ($classNames as $className) {
       if (!$this->isClassNameSupported($className)) {
            return false;
       }
}

return true;
```

In order to support
```
$query = rand(0, 1) ? $query1 : $query2;
$product = $queryBus->dispatch($query);
```

bnowak replied:
:+1: adjusted

bnowak committed:
added union test-cases

bnowak committed:
support for handling union types

bnowak commented:
@VincentLanglet thank you for your reply and review. I've addressed your comments and also added support for union types in classes which use `HandleTrait` directly (not only on their wrappers/query buses).

Please take a look and re-review.

Will that change be automatically merged into 2.x branch as well when approved? :crossed_fingers:

VincentLanglet commented:
Oh, didn't see you were targeting the 1.4.x branch.

You should target the 2.0.x instead ; there is no feature added to the version 1.

bnowak committed:
messenger params injected as service argument

VincentLanglet commented on tests/Type/Symfony/data/messenger_handle_trait.php:
```diff
@@ -50,7 +50,69 @@ public function __invoke()
+class QueryBus {
+    use HandleTrait;
+
+    public function dispatch(object $query): mixed
```
To avoid the failure on php 7.4 I think you have to replace `: mixed` by a phpdoc `@return mixed` (everywhere in this file)

bnowak replied:
thanks for advice. I've tried that, but with just adding phpdoc it didn't change anything. It started working when I simply removed native return type (it's mixed by default).

Now it should work (I've tested locally on PHP 7.4 & 8.4), but now one CI job failed (on  github.com authentication). I don't see button for retrying it (perhaps I don't have such permissions).

@VincentLanglet could you take a look and rerun CI jobs, please?

VincentLanglet replied:
I restarted the Ci it's ok

bnowak committed:
fixed return mixed type wrong inference

bnowak commented:
@VincentLanglet rebased to 2.0.x branch, but there's still issue with failed CI job (mentioned [here](https://github.com/phpstan/phpstan-symfony/pull/423#discussion_r2822914790)).

Could you take a look on that and re-review once again? I hope, I already covered everything in this PR :pray:

VincentLanglet commented:
Since you initially requested changes @ondrejmirtes, your review is required here :)

bnowak commented:
Thank you @VincentLanglet!

@ondrejmirtes kindly request to take a look at that🙏

ondrejmirtes commented:
Thank you!
